### DISCUSSION TITLE
[help] How to make autocompletion works in blade?

### DISCUSSION THREAD

**Discussion by feryardiant** [EmranMR/tree-sitter-blade#72] [Q&A] [answered]
First of all, thanks for the amazing work in this plugin. I've spent whole day today setting up my [neovim config](https://github.com/feryardiant/dotfiles/pull/17) to work with blade files.

https://github.com/user-attachments/assets/18525d57-82b0-4c5d-9c8e-07d9207355cf

Everything seems working just fine, until I realize that :

- Only directives that defined in [`grammar.json`](https://github.com/EmranMR/tree-sitter-blade/blob/335b2a44b4cdd9446f1c01434226267a61851405/src/grammar.json) is highlighted, other than that (directive from third-party package, e.g: inertia) is not.
- The auto-completion (me using `intelephense`) is not working.
- There's `ERROR` in the `InspectTree` on directive' param.

Can anyone guide me regarding above points? Thank you in advance.

My config: https://github.com/feryardiant/dotfiles/pull/17

Best regards,
Fery

**Comment by EmranMR** [comment]
Hi @feryardiant, and thanks for the kind words!

#### Inertia
- I am not an Inertia user, can you direct me where I can find all the directives for inertia, or list them here, and I will add them to the grammar.
- The errors are probably because of the inertia directives in your case, which will be fixed once I know where I can find them all.

#### Autocompletion.
- This is a tricky one, intelephense LSP [does not recognise the blade files](https://github.com/bmewburn/vscode-intelephense/issues/93). 
- I can confirm the emmet LSP does work in blade files, but you need to adjust your config files. I successfully got it working for Helix. But that is just for the HTML stuff
- For Nova, they have API for userland completion and LSP, so I just used the API to create the extension and implement other language functions. 

##### Workaround for NeoVim?
- well, I am not sure if there are extensions that provide autocompletion
- I have the same issue in Helix, but yet to try out the snippet LSP, and just basically make a snippet library. Of course these probably would not be context aware and dynamic like intelephense... 

Hopefully the new extension laravel is developing for vscode is an actual LSP which would resolve it practically for all editors!

**Reply by feryardiant** [reply]
> * can you direct me where I can find all the directives for inertia, or list them here, and I will add them to the grammar

- `@inertia` and `@inertiaHead` are from [`inertiajs/inertia-laravel`](https://github.com/inertiajs/inertia-laravel/blob/ddf47e999dcdabd2f370e5b9772a7449f4a81729/src/ServiceProvider.php#L55-L58) package
- `@routes` is from [`tighten/ziggy`](https://github.com/tighten/ziggy/blob/a8a5984adfc1b9c987fe0a6b6e7db4555ab2cf16/src/ZiggyServiceProvider.php#L31) package

> * The errors are probably because of the inertia directives in your case

I'm not sure that was the case, as you can see when I click the "error" line on the `InspectTree` side, it'll highlight the line on the blade file side which is on `@vite()` directive.

![Screenshot 2024-09-17 at 20 59 26](https://github.com/user-attachments/assets/177ac457-772e-43af-b548-a7d74698eb8b)

> * This is a tricky one, intelephense LSP

Indeed, based on [this](https://github.com/calebdw/dotfiles/blob/45144348111937fbec4fdb29dad33300471d5fe5/.config/nvim/lua/calebdw/plugins/lsp.lua#L210-L218) config by @calebdw, I made [this](https://github.com/feryardiant/dotfiles/blob/864db3d6fe94326e23c5167703eeb95c55c94807/nvim/lua/plugins/lsp.lua#L224-L229) and hopefully it could work but it's not 😅

Even though I can see that `intelephense` is listed when I run `:LspInfo` on that blade file.

![Screenshot 2024-09-17 at 21 05 33](https://github.com/user-attachments/assets/1b71e11f-aaae-4459-9b90-5ed14146188d)

> Hopefully the new extension laravel is developing for vscode is an actual LSP which would resolve it practically for all editors!

Yes! and I hope they have free tier like what `intelephense` does.

**Reply by calebdw** [reply]
Errors are not a concern in and of themselves---you need to determine what language they are occurring in (toggle the languages in the tree). Errors in injections are expected

**Reply by calebdw** [reply]
Intelephense likely isn't going to work for blade files, even with the php injections because the language server has no context to work with

**Reply by feryardiant** [reply]
> Errors are not a concern in and of themselves---you need to determine what language they are occurring in (toggle the languages in the tree). Errors in injections are expected

Yes, the error isn't bothering me that much, but if there's a way to "fix" it, that would be better. 😁

> Intelephense likely isn't going to work for blade files, even with the php injections because the language server has no context to work with

That's unfortunate 🥲

**Reply by calebdw** [reply]
Here's something you might be interested in:

https://github.com/ricardoramirezr/blade-nav.nvim

**Comment by feryardiant** [comment]
I think I've already got the answer for the autocompletion and the error message, but not the grammar.

@EmranMR did you have any plan to add supports for directives from third-party packages?

**Reply by EmranMR** [reply, marked as answer]
Hi @feryardiant I just pushed the new version with Inertia and Ziggy directives support.
> did you have any plan to add supports for directives from third-party packages?

The [custom if statements](#3) are already supported as long as the third party package, follows the Laravel convention. If there are none-standard directives, they need to be defined by the grammar. For example, Livewire 3 has some directives that are not hardcoded and already supported. 

For `keyword` style directives, they all need to be defined manually. There are actually not that many keywords as far as I know,   and thankfully it is super easy to extend the grammar. However, dynamic support is fairly impossible, to give you an example, you can not simply use RegEx to say `@whatever` should be parsed as directive, this will lead to issues with`test@example.com`. JavaScript grammar definitions for `tree-sitter` itself does not support `lookahead` and `lookbehinds`.
